database/ffldb/db.go

```diff
@@ -2080,5 +2246,42 @@ func openDB(dbPath string, network wire.BitcoinNet, create bool) (database.DB, e
 
 	// Perform any reconciliation needed between the block and metadata as
 	// well as database initialization, if needed.
-	return reconcileDB(pdb, create)
+	if _, err := reconcileDB(pdb, create); err != nil {
+		return nil, err
+	}
+
+	if !create {
+		// Run all required data migrations
+		if err := migrateBlockIndex(pdb); err != nil {
+			return nil, err
```

Comment:
While I was going back and forth between this branch and another, I added some code to delete the migration key and new buckets, but I noticed some strange behavior where they weren't actually being stored to the database when there was an error, despite the underlying transactions having completed successfully.

After thinking about it for a bit, I realized the problem is that the database needs to be closed at any point after the `pdb` is created above when there are any errors before returning or it won't close the underlying cache and flush to disk.  This is true for the `reconcileDB` error path above, this one, and the lookup path below.  The fact it wasn't doing it before for the `reconcileDB` error return was also a bug.

```Go
		// Run all required data migrations.
		if err := migrateBlockIndex(pdb); err != nil {
			// Ignore error on close since no further recovery could
			// be done anyways.
			_ = pdb.Close()
			return nil, err
		}
```